storage/replica.go

```diff
@@ -519,32 +522,56 @@ func loadGCThreshold(eng engine.Reader, rangeID roachpb.RangeID) (roachpb.Timest
 
 // newNotLeaderError returns a NotLeaderError initialized with the
 // replica for the holder (if any) of the given lease.
-// r.mu must not be held.
-func (r *Replica) newNotLeaderError(l *roachpb.Lease, originStoreID roachpb.StoreID) error {
-	r.mu.Lock()
-	defer r.mu.Unlock()
-	return r.newNotLeaderErrorLocked(l, originStoreID)
-}
-
-func (r *Replica) newNotLeaderErrorLocked(l *roachpb.Lease, originStoreID roachpb.StoreID) error {
-	desc := r.mu.desc
+func (r *Replica) newNotLeaderError(
+	l *roachpb.Lease, originStoreID roachpb.StoreID, rangeDesc *roachpb.RangeDescriptor,
+) error {
 	err := &roachpb.NotLeaderError{}
 	if l != nil && l.Replica.ReplicaID != 0 {
+		desc := rangeDesc
+
 		err.RangeID = r.RangeID
 		_, err.Replica = desc.FindReplica(originStoreID)
 		_, err.Leader = desc.FindReplica(l.Replica.StoreID)
 	}
 	return err
 }
 
-// redirectOnOrAcquireLeaderLease checks whether this replica has the
-// leader lease at the specified timestamp. If it does, returns
-// success. If another replica currently holds the lease, redirects by
-// returning NotLeaderError. If the lease is expired, a renewal is
-// synchronously requested. This method uses the leader lease mutex
-// to guarantee only one request to grant the lease is pending.
-// Leases are eagerly renewed when a request with a timestamp close to
-// the beginning of the stasis period is served.
+// checkLeaseLocked checks that the replica owns the lease at the specified
+// timestamp.
+// It returns true if the current lease covers the timestamp, false if it
+// doesn't. If the lease is not owned, a NotLeaderError is returned.
+// The lease is also extended if owned and almost expiring.
+func (r *Replica) checkLeaseLocked(timestamp roachpb.Timestamp) (bool, *roachpb.Error) {
```

Comment:
`check` does not suggest taking action when there is no lease.

I don't see the point in factoring this out into its own method. There are enough methods with `lease` in their names to be confusing, and it's only used from one of the other ones from which you extracted it.